src/python/pants/backend/python/providers/pyenv/rules.py

```diff
@@ -0,0 +1,338 @@
+from pants.engine.internals.native_engine import Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.internals.synthetic_targets import SyntheticAddressMaps, SyntheticTargetsRequest
+from pants.engine.internals.target_adaptor import TargetAdaptor
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessCacheScope, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.util.docutil import bin_name
+from pants.util.frozendict import FrozenDict
+from pants.util.meta import classproperty
+from pants.util.strutil import softwrap
+
+PYENV_NAMED_CACHE = ".pyenv"
+PYENV_APPEND_ONLY_CACHES = FrozenDict({"pyenv": PYENV_NAMED_CACHE})
+
+
+class PyenvSubsystem(TemplatedExternalTool):
+    options_scope = "pyenv"
+    name = "pyenv"
+    help = softwrap(
+        f"""
+        A subsystem for pyenv-provided-Python (https://github.com/pyenv/pyenv).
+
+        Enabling this subsystem will switch Pants from trying to find an appropriate Python on your
+        system to using pyenv to install the correct Python(s).
+
+        The Pythons provided by Pyenv will be used to run any "user" code (your Python code as well
+        as any Python-based tools you use, like black or pylint). The Pythons are also read-only to
+        ensure they remain hermetic across runs of different tools and code.
+
+        The Pythons themselves are stored in your `named_caches_dir`: https://www.pantsbuild.org/docs/reference-global#named_caches_dir
+        under `pyenv/versions/<version>`. Wiping the relevant version directory (with `sudo rm -rf`)
+        will force a re-install of Python. This may be necessary after changing something about the
+        underlying system which changes the compiled Python, such as installing an
+        optional-at-build-time dependency like `liblzma-dev` (which is used for the optional module
+        `lzma`).
+
+        By default, the subsystem does not pass any optimization flags to the Python compilation
+        process. Doing so would increase the time it takes to install a single Python by about an
+        order of magnitude (E.g. ~2.5 minutes to ~26 minutes).
+
+        If you wish to customize the pyenv installation of a Python, a synthetic target is exposed
+        at the root of your repo which is runnable. This target can be run with the relevant
+        environment variables set to enable optimizations. You will need to wipe the specific
+        version directory if Python was already installed. Example:
+
+            sudo rm -rf <named_caches_dir>/pyenv/versions/<specific_version>
+            # env vars from https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#building-for-maximum-performance
+            PYTHON_CONFIGURE_OPTS='--enable-optimizations --with-lto' {bin_name()} run :pants-pyenv-install -- 3.10
```

Comment:
If we expose this through options it's the same end result. Customized python builds...

And this way is the only "middle ground" I could come up with to make having some (uncacheable) build step take 30 minutes. Make the user "own" running that.

Reply:
Well maybe I'm just viewing this from my perspective, but I know my engineers wouldn't wait for a 30 minute process. And they'd bug me constantly about it. And kill it. Meaning they'd have to restart.

So I can't put it in pants.toml. 

And putting it there feels weird because if you change the option we don't invalidate existing pythons either. (Same for .pants.rc)

Reply:
All in all, the end goal is providing Python. That's the 80% case.

For the 20%, let's give them a back door. But it doesn't have to be of the same elk as the front door 😉

Reply:
> So I can't put it in pants.toml.

This is true for optimization flags, but not true for the other flags you might want to pass to a `pyenv` install: particularly `CONFIGURE_OPTS`.

> And putting it there feels weird because if you change the option we don't invalidate existing pythons either. (Same for .pants.rc)

We could? Fingerprinting the options could be done by embedding them in the installation script, and taking its digest, for example.

Reply:
Sure anything is possible. Getting back to it though, the build is already non-hermetic and we still need a way to allow users to install Python out-of-band or suffer a 30-minute build step.

Reply:
OK I think I've decided to have this functionality be in its own sub-plugin of the `pyenv` provider plugin. So it's "off" by default.